sql/select.go

```diff
@@ -85,3 +99,523 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
+		} else {
+			// "a BETWEEN b AND c" -> "a >= b AND a <= c"
+			m.analyzeExpr(&parser.AndExpr{
+				Left: &parser.ComparisonExpr{
+					Operator: parser.GE,
+					Left:     t.Left,
+					Right:    t.From,
+				},
+				Right: &parser.ComparisonExpr{
+					Operator: parser.LE,
+					Left:     t.Left,
+					Right:    t.To,
+				},
+			})
+		}
+	}
+}
+
+// analyzeComparison analyzes the comparison expression, restricting the start
+// and end info in indexInfo.elems if possible.
+func (m qvalueRangeMap) analyzeComparisonExpr(node *parser.ComparisonExpr) {
+	op := node.Operator
+	switch op {
+	case parser.EQ, parser.LT, parser.LE, parser.GT, parser.GE:
+		break
+
+	default:
+		// TODO(pmattis): For parser.LIKE we could extract the constant prefix and
+		// treat as a range restriction.
+		return
+	}
+
+	qval, ok := node.Left.(*qvalue)
+	var constExpr parser.Expr
+	if ok {
+		if !isConst(node.Right) {
+			return
+		}
+		constExpr = node.Right
+		// qval <op> constant
+	} else {
+		qval, ok = node.Right.(*qvalue)
+		if !ok {
+			return
+		}
+		if !isConst(node.Left) {
+			return
+		}
+		constExpr = node.Left
+		// constant <op> qval: invert the operation.
+		switch op {
+		case parser.LT:
+			op = parser.GT
+		case parser.LE:
+			op = parser.GE
+		case parser.GT:
+			op = parser.LT
+		case parser.GE:
+			op = parser.LE
+		}
+	}
+
+	// Evaluate the constant expression.
+	datum, err := parser.EvalExpr(constExpr)
+	if err != nil {
+		return
+	}
+
+	// Ensure the resulting datum matches the column type.
+	_, err = convertDatum(qval.col, datum)
+	if err != nil {
+		return
+	}
+
+	if log.V(2) {
+		log.Infof("analyzeComparisonExpr: %s %s %s",
+			qval.col.Name, op, datum)
+	}
+
+	r := m.getRange(qval.col.ID)
+
+	// Restrict the start element.
+	switch startOp := op; startOp {
+	case parser.GE:
+		startOp = parser.EQ
+		fallthrough
+	case parser.EQ, parser.GT:
+		r.start.intersectStart(qvalueInfo{datum, startOp})
+	}
+
+	// Restrict the end element.
+	switch endOp := op; endOp {
+	case parser.LE:
+		endOp = parser.EQ
+		fallthrough
+	case parser.EQ, parser.LT:
+		r.end.intersectEnd(qvalueInfo{datum, endOp})
+	}
+}
+
+func (m qvalueRangeMap) getRange(colID ColumnID) *qvalueRange {
+	r := m[colID]
+	if r == nil {
+		r = &qvalueRange{}
+		m[colID] = r
+	}
+	return r
+}
+
+type indexInfo struct {
+	desc  *TableDescriptor
+	index *IndexDescriptor
+	start []qvalueInfo
+	end   []qvalueInfo
+	cost  float64
+}
+
+// analyzeRanges analyzes the scanNode and range map to determine the cost of
+// using the index.
+func (v *indexInfo) analyzeRanges(s *scanNode, m qvalueRangeMap) {
+	if !v.isCoveringIndex(s.qvals) {
+		// TODO(pmattis): Support non-coverying indexes.
+		v.cost = math.MaxFloat64
+		return
+	}
+
+	v.makeStartInfo(m)
+	v.makeEndInfo(m)
+
+	// The base cost is the number of keys per row.
+	if v.index == &v.desc.PrimaryIndex {
+		// The primary index contains 1 key per column plus the sentinel key per
+		// row.
+		v.cost = float64(1 + len(v.desc.Columns))
```

Comment:
shouldn't that be `1 + len(v.desc.Columns) - len(v.desc.PrimaryIndex.ColumnIDs)`?